src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,265 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations.FlowAnalysis;
+using static Microsoft.CodeQuality.Analyzers.MicrosoftCodeQualityAnalyzersResources;
+using static Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations.AvoidMultipleEnumerationsHelpers;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations
+{
+    internal abstract partial class AvoidMultipleEnumerations : DiagnosticAnalyzer
+    {
+        private const string RuleId = "CA1851";
+
+        private static readonly DiagnosticDescriptor MultipleEnumerableDescriptor = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(AvoidMultipleEnumerationsTitle)),
+            CreateLocalizableResourceString(nameof(AvoidMultipleEnumerationsMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.Disabled,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(MultipleEnumerableDescriptor);
+
+        /// <summary>
+        /// Additional types that has the ability to defer enumeration.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_additionalDeferredTypes =
+            ImmutableArray.Create(WellKnownTypeNames.SystemLinqIOrderedEnumerable1);
+
+        /// <summary>
+        /// All the immutable collections that have a conversion method from IEnumerable.
+        /// </summary>
+        private static readonly ImmutableArray<(string typeName, string methodName)> s_immutableCollectionsTypeNamesAndConvensionMethods = ImmutableArray.Create(
+            ("System.Collections.Immutable.ImmutableArray", nameof(ImmutableArray.ToImmutableArray)),
+            ("System.Collections.Immutable.ImmutableDictionary", nameof(ImmutableDictionary.ToImmutableDictionary)),
+            ("System.Collections.Immutable.ImmutableHashSet", nameof(ImmutableHashSet.ToImmutableHashSet)),
+            ("System.Collections.Immutable.ImmutableList", nameof(ImmutableList.ToImmutableList)),
+            ("System.Collections.Immutable.ImmutableSortedDictionary", nameof(ImmutableSortedDictionary.ToImmutableSortedDictionary)),
+            ("System.Collections.Immutable.ImmutableSortedSet", nameof(ImmutableSortedSet.ToImmutableSortedSet)));
```

Comment:
These are usually put in `WellKnownTypeNames`.

`System.Collections.Immutable.ImmutableDictionary` and `System.Collections.Immutable.ImmutableHashSet` are already there.

https://github.com/dotnet/roslyn-analyzers/blob/c1736d0da9501e135f0410b5a831528f699d6e60/src/Utilities/Compiler/WellKnownTypeNames.cs#L143-L145